### ISSUE TITLE
Use of reusable workflows for ci

### ISSUE THREAD

**Issue by mcbarton** [compiler-research/CppInterOp#279]
Now that the ci in the repos are approximately the same, we are now in a position to start replacing it with reusable workflows from a central repo. @vgvassilev could you make a repo for these workflows, and then change the settings of that repo according to what is described in the below link

https://github.blog/2022-02-10-using-reusable-workflows-github-actions/

After this has been done I will begin the process of replacing parts of the ci in this repo (and others) bit by bit with the workflows in this repo. That way the ci is all the repos should then stay upto date with each other.

**Comment by vgvassilev** [comment]
Can you propose a repo name?

**Comment by mcbarton** [comment]
> Can you propose a repo name?

I'd just go with something simple like 'Reusable-Workflows' or 'Reusable-Actions' . That way its clear to anyone who clicks on the organisation what it contains.

**Comment by maximusron** [comment]
Looks cool! Does this allow us write to a single workflow for the CI that can be used as the main workflow for the three cppyy repo and as a subworkflow within InterOp?

**Comment by mcbarton** [comment]
> Looks cool! Does this allow us write to a single workflow for the CI that can be used as the main workflow for the three cppyy repo and as a subworkflow within InterOp? That would be amazing

You'll be able to write to worflows in this new repo, and all the repos will use it accordingly. There will be no need in future to update it in all the repos separately. The cppyy/CppInterOp repos will call the workflows from this new repo.

**Comment by maximusron** [comment]
Yep I understand that, my question was more about the way we pull these workflows and how we choose to discretize their parts. For example up until the Build and Test CppInterOp will differ in cppyy repos from the CppInterOp repos. The cppyy installation + testing will remain common. So we need to break it into reusable parts that make the most sense. For example 
<img width="427" alt="image" src="https://github.com/compiler-research/CppInterOp/assets/75925957/6abeb72b-bee7-4e30-bebf-21a6ca865758">

**Comment by mcbarton** [comment]
> Yep I understand that, my question was more about the way we pull these workflows and how we choose to discretize their parts. For example up until the Build and Test CppInterOp will differ in cppyy repos from the CppInterOp repos. The cppyy installation + testing will remain common. So we need to break it into reusable parts that make the most sense. For example <img alt="image" width="427" src="https://private-user-images.githubusercontent.com/75925957/330396642-6abeb72b-bee7-4e30-bebf-21a6ca865758.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU2ODQ3NjIsIm5iZiI6MTcxNTY4NDQ2MiwicGF0aCI6Ii83NTkyNTk1Ny8zMzAzOTY2NDItNmFiZWI3MmItYmVlNy00ZTMwLWJlYmYtMjFhNmNhODY1NzU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE0VDExMDEwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg4YThlYmVkNTFjOTNkMGY0ZWEyOGU4ZjI0ODkyZGVmYTI1NDViMzA1Y2Y0YjJiMGIzNGExMzA5NGJjMTc5NWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.YTko97nWelmfKlmOPv8iwhn0msImADQxE6GTCKzo9jU">

@maximusron How you would like to see the workflow broken down into smaller chunks its probably more of a discussion for you and @vgvassilev to discuss. In the case where you want a Valgrind check, documentation, coverage, etc this can be done with workflow inputs. I'll try and implement the way you choose, and cc you if I run into any difficulties.

**Comment by vgvassilev** [comment]
> > Can you propose a repo name?
> 
> I'd just go with something simple like 'Reusable-Workflows' or 'Reusable-Actions' . That way its clear to anyone who clicks on the organisation what it contains.

Do we envision anybody else beyond our org to use these actions? If so, we need to have some sort of prefix?

**Comment by mcbarton** [comment]
![image](https://github.com/compiler-research/CppInterOp/assets/150042563/637757d8-4caf-408e-a8c0-581f514e6cad)

 @vgvassilev This part of the link I put in the issue says you set it so it cannot be used outside the organisation.

**Comment by vgvassilev** [comment]
I meant if we expect the workflows to be generic enough so that somebody else outside of our org to find them useful and possibly use them.

**Comment by mcbarton** [comment]
> I meant if we expect the workflows to be generic enough so that somebody else outside of our org to find them useful and possibly use them.

@vgvassilev There may be some workflows generic enough for others to use them (for example I can probably make the llvm cache one generic enough). I suspect the majority at least at first will specific to the organisation. I don't know what name to put for the repo in these circumstances.

**Comment by vgvassilev** [comment]
@mcbarton, https://github.com/compiler-research/ci-workflows

I've made you maintainer and you should be able to do almost everything.
